### ISSUE TITLE
Resolve requirements

### ISSUE THREAD

**Issue by garrettwrong** [ComputationalCryoEM/ASPIRE-Python#192]
There are some discrepancies between `requirements.txt` and `setup.py`.  For example, `pytest`. While it is technically correct to omit from `install_requires` the expectation is usually that it would instead be placed in a `test_requires` (which we don't do/have), there is not a corollary requirements.txt (usually just the one file for those who use it).  I feel users and developers might have the expectation that if they pip install the code, and we generally suggest running `pytest` that it would work out of the box. 

I actually don't know why we have both `install_requires` in `setup.py` and `requirements.txt`, it is confusing, even if they were identical.  Probably the idea was we're going to support both conda and pip with one requirements file, and in that case `setup.py` can just parse the file, instead of duplicating. That would be a first pass fix I can do pretty quickly.

For specialized extensions (`extra_requires`), again I do not think is an exact parallel between conda and `setup.py`, though we could have two requirements, perhaps as an additional `requirements_gpu.txt` that would load cufinufft, cupy and/or pycuda etc. IE the stuff that would potentially break a normal install. This could be read into `setup.py` extras section; again to avoid dupes. Or we can just drop the requirements files all together. I'm not exactly sure why they exist here... (I think conda can just pip install the package?)

@vineetbansal , we talked about this once, do you recall any other reason? Thanks

**Comment by janden** [comment]
I was the one that asked a requirements.txt file to be added. I wasn't familiar enough with setuptools to figure out how they overlap. If there's a clean way to make setup.py install the required packages (whether test, extra, or otherwise), I'm totally fine gutting the requirements file (as long as Conda can still access the required packages through setup.py).

**Comment by garrettwrong** [comment]
One option is from within setup.py, just read the file, split and strip it into a list. You'll see a lot of codes do that.

The other is to just try to let pip manage it within conda, dropping the file as you suggest.  I don't think that is a problem, but need to check; that is probably best if possible.

Neither of those to are much work.

**Comment by vineetbansal** [comment]
I think I added a `requirements.txt` for developers (of which @janden was one) who wish to set up a virtualenv to work on the code (not necessarily to just install it away to start using it as an end user). Everything does work in such a pypi-only scenario as we discovered, but `environment.yml` is a rather blunt tool to avoid dependency surprises, so it's nice to have it around as well. The `environment.yml` referring to `requirements.txt` was supposed to support both these scenarios without redundant info.

But I agree that it's equally easy for a developer who wants to avoid conda to install the package in develop mode, in which case `install_requires` can suffice. If there's a way for `install_requires` to refer to this `requirements.txt` file, that to me looks to be the proper way. One can reasonably disagree and think that'll make the confusion even worse - in that case it's probably best to have all dependencies explicitly specified in `install_requires`, and a separate standalone `environment.yml` for conda users.

I would probably avoid reading `environment.yml` from within `setup.py`, since that to me is an inversion of scopes. (**Edit**: when I think about it, it also seems to me to be the wrong thing to do, since identical version numbers across conda <-> pypi do not necessarily imply identical or even compatible codebases, unless things have reached a happy convergence on that front). 

I agree about the `test_requires` - I wasn't aware of that, but that's good to know!

**Comment by garrettwrong** [comment]
Gotcha, okay, venv makes sense. 

Btw, I meant we can read `requirements.txt` from `setup.py`... this is common, fwiw. I have no intention of reading `environment.yml`, you are correct, so we're good on that front :)

Cool thanks for the backround. I didn't even consider the venv use case.  I think we can make everything consistent, no worries.

**Comment by janden** [comment]
> we can read `requirements.txt` from `setup.py`

I think that makes sense then. This way we can support both conda and setuptools since they both read from `requirements.txt`?

**Comment by garrettwrong** [comment]
I don't think conda is a problem (conda should just be able to use pip which uses `setup.py`). Considering it may have been for virtualenv, I can sort of understand... The case Vineet described is someone wanting to install the stack of depends into a venv, w/o actually installing the corresponding aspire yet... they would not use setuptools or conda. I'm not sure how that case got precedence over pip install, but that's what this issue is going to fix....

Translated slightly, yes, virtualenv and setup.py could both read `requirements.txt`.  The only thing the text file doesn't do is allow us to call out separate special dependencies, like for GPU which would break some users, so you'd end up with two (or more files), which is sort of ugly, but would work okay.

To explain:

pip has a really nice feature where you can define optional extras, so a regular user might do:

```
#reg user
pip install aspire
# gpu user
pip install aspire[gpu]
# developers that wants extra tools like (isort, twine, pep tools etc)
pip install aspire[dev]
```

Internally setuptools actually separates this for `install_requires` `setup_requires` `tests_requires` then gives you `extras` which is user definable dict for custom stuff like above.  Not that we need all of it, maybe just install and extras, but for context.  So to do what pip does in a way that supports venv, you would have separate text files for those cases (ignore conda, conda should probably use pip insidee, I don't think conda is actually relevant the to issue beyond that).

**Comment by janden** [comment]
> The case Vineet described is someone wanting to install the stack of depends into a venv, w/o actually installing the corresponding aspire yet... they would not use setuptools or conda. I'm not sure how that case got precedence over pip install, but that's what this issue is going to fix....

This may simply stem from me not understanding how setuptools. What I would do is `pip install -r requirements`, then `python setup.py develop`. From what you're saying, the latter should be enough. With these optional extras, we're also getting more flexibility, so if we need to ditch `requirements.txt` to do that, I'm fine with that.

**Comment by garrettwrong** [comment]
Okay, cool, the corner case was a little odd. This should work out great, I'm glad we all talked about it. I'll try to put a PR together tomorrow morning.

**Comment by garrettwrong** [comment]
This should have been closed by #197
